pkg/server/skydb/access.go

```diff
@@ -126,3 +130,192 @@ func (acl RecordACL) Accessible(userinfo *UserInfo, level RecordACLLevel) bool {
+}
+
+// Accessible returns true when the access mode is allowed access
+func (acl FieldACL) Accessible(
+	userinfo *UserInfo,
+	record *Record,
+	recordType string,
+	field string,
+	mode FieldAccessMode,
+) bool {
+	if acl.wildcardRecordType.Accessible(userinfo, record, recordType, field, mode) {
+		return true
+	}
+	if list, ok := acl.recordTypes[recordType]; ok {
+		return list.Accessible(userinfo, record, recordType, field, mode)
+	}
+	return false
+}
+
+// FieldACLEntryList contains a list of field ACL entries
+type FieldACLEntryList []FieldACLEntry
+
+// Accessible returns true when the access mode is allowed access
+func (list FieldACLEntryList) Accessible(
+	userinfo *UserInfo,
+	record *Record,
+	recordType string,
+	field string,
+	mode FieldAccessMode,
+) bool {
+	for _, entry := range list {
+		if entry.Accessible(userinfo, record, recordType, field, mode) {
+			return true
+		}
+	}
+	return false
+}
+
+func (list FieldACLEntryList) findDefaultEntry() *FieldACLEntry {
+	for _, entry := range list {
+		if entry.RecordType == WildcardRecordType &&
+			entry.RecordField == WildcardRecordField &&
+			entry.UserRole == "_public" {
+			return &entry
+		}
+	}
+	return nil
+}
+
+func (list FieldACLEntryList) Len() int      { return len(list) }
+func (list FieldACLEntryList) Swap(i, j int) { list[i], list[j] = list[j], list[i] }
+func (list FieldACLEntryList) Less(i, j int) bool {
+	// compare is similar to strings.Compare except that specified wildcard
+	// string will be less than non-wildcard string.
+	compare := func(a, b, wildcard string) int {
+		if a == wildcard && b != wildcard {
+			return -1
+		} else if b == wildcard && a != wildcard {
+			return 1
+		}
+		return strings.Compare(a, b)
+	}
+
+	result := compare(list[i].RecordType, list[j].RecordType, WildcardRecordType)
+	if result != 0 {
+		return result < 0
+	}
+
+	result = compare(list[i].RecordField, list[j].RecordField, WildcardRecordField)
+	if result != 0 {
+		return result < 0
+	}
+
+	return strings.Compare(list[i].UserRole, list[j].UserRole) < 0
+}
+
+// FieldACLEntry contains a single field ACL entry
+type FieldACLEntry struct {
+	RecordType   string
+	RecordField  string
+	UserRole     string
+	Writable     bool
+	Readable     bool
+	Comparable   bool
+	Discoverable bool
+}
+
+// Accessible returns true when the access mode is allowed access
+func (entry FieldACLEntry) Accessible(
+	userinfo *UserInfo,
+	record *Record,
+	recordType string,
+	field string,
+	mode FieldAccessMode,
+) bool {
+	if (entry.RecordType != recordType && entry.RecordType != WildcardRecordType) ||
+		(entry.RecordField != field && entry.RecordField != WildcardRecordField) {
+		return false
+	}
+
+	// TODO: Check access here
+
+	return (mode == ReadFieldAccessMode && entry.Readable) ||
+		(mode == WriteFieldAccessMode && entry.Writable) ||
+		(mode == CompareFieldAccessMode && entry.Comparable) ||
+		(mode == DiscoverFieldAccessMode && entry.Discoverable)
```

Comment:
In the [feature proposal](https://github.com/SkygearIO/features/blob/master/features/48-user-profile-discovery/field-based-acl.md#access-control-entity-ace), the ACL entity only has 2 fields representing accessibility level and discovery level, which implies that `ReadWrite` is superset of `Readable` and `Queryable` is superset of `Discoverable`. 

But in the [API changes section of the proposal](https://github.com/SkygearIO/features/blob/master/features/48-user-profile-discovery/field-based-acl.md#changes-on-api-at-skygear-server), `ReadWrite` is separated into `readable` and `writeable`. So that means that assumption is ensure on SDK and Skygear server will not handle that situation. 

Do you agree?